### ISSUE TITLE
False positives for CloudLinux

### ISSUE THREAD

**Issue by edie-etoile** [speed47/spectre-meltdown-checker#147]
Hello,
We ran your checker on CloudLinux Kernel: 3.10.0-714.10.2.lve1.5.12.el6h.x86_64

And is shows vulnerabilities. CloudLinux says:

```
Daniil Ivanov (CloudLinux)Feb 15, 23:30 EET Hello
It seems that your checker provides inappropriate information. Easy check shows that:

# [root@host ~]# rpm -q --changelog kernel \| grep 2017-575
# - Added patches for Meltdown and Spectre attacks (CVE-2017-5753, CVE-2017-5715,  CVE-2017-5754)
# - Added patches for Meltdown and Spectre attacks (CVE-2017-5753, CVE-2017-5715,  CVE-2017-5754)

I suggest you write this bug down to the aforementioned developer for further fix. If you have any other issues or questions - let me know
--
```

See our attached output from your script.
[1802151619-vulnerable-still.txt](https://github.com/speed47/spectre-meltdown-checker/files/1729489/1802151619-vulnerable-still.txt)

**Comment by knweiss** [comment]
@edie-etoile (I’m not the developer) Your kernel seems to have the mitigations/fixes. However, **these** kernel patches **require** uptodate CPU microcode (with support for IBPB and IBRS) to be able to mitigate Spectre variant 2. Without these mandatory microcode features the kernel is still vulnerable.

Unfortunately, Intel recalled these microcode updates due to stability problems and is now step for step releasing fixed versions. See the [Intel Microcode Guidance](https://newsroom.intel.com/wp-content/uploads/sites/11/2018/02/microcode-update-guidance.pdf) for details. At the moment the microcode of your Xeon Ivy Bridge E3 is in „Pre-Beta“ status.

(The other Spectre variant 2 mitigation alternative that doesn‘t require a microcode update (at least on your cpu) is to compile the kernel with the retpoline feature (that your kernel doesn’t have!). However, this would require a compiler with retpoline support...)

Long story short: From my point of view the output is correct.

**Comment by speed47** [comment]
Thanks also my understanding (thanks @knweiss !)
I just released v0.35. The message for variant2 is now more clear when IBRS is supported in kernel but the CPU microcode doesn't support it

**Comment by edie-etoile** [comment]
This is the CloudLinux response:

```

Vladislav Dagaev (CloudLinux)Feb 16, 07:10 EET Hello.Checker produces a false-negative result. The checker is counting 'lfence'instructions in kernel image on disk, not in memory. KernelCare appliespatches only to the in-memory kernel image. Thus, the checker is counting'lfence' instruction only in unpatched image.Thank you
--



```

**Comment by speed47** [comment]
Indeed, I'm almost sure I can't access the kernel image directly in memory, to detect live patching (done by tools such as KernelCare).
As I suppose CloudLinux will (or already has) release(d) a patched kernel image (on disk), you can use the checker to verify this image.

However CloudLinux support seems to talk about the variant1 detection, but in the full output you pasted above, the variant1 mitigation is correctly detected (in your on-disk kernel image). They *might* have patched variant2 in-memory, and if it's the case I can't detect it, but their release notes most probably explicit this part.

**Comment by edie-etoile** [comment]
Your tool isn't capable of testing CloudLinux in memory patches.
